### PR TITLE
Fix bitmap font error when parsing '-'

### LINKED ISSUE #82880: Monospace Image Font does not handle '-' range correctly
### Godot version

4.2.dev5.mono

### System information

Godot v4.2.dev5.mono - Windows 10.0.19045 - Vulkan (Mobile) - integrated Intel(R) Iris(R) Xe Graphics (Intel Corporation; 30.0.100.9836) - 11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz (8 Threads)

### Issue description

When importing an image as a monospace image font, the parsing of character ranges does not work properly. This is how each range is supposed to be parsed, according to [this page](https://docs.godotengine.org/en/stable/tutorials/ui/gui_using_fonts.html#bitmap-fonts) in the documentation:
> The Character Ranges option is an array that maps each position on the image (in tile coordinates, not pixels). The font atlas is traversed from left to right and top to bottom. Characters can be specified with decimal numbers (127), hexadecimal numbers (0x007f) or between single quotes ('~'). Ranges can be specified with a hyphen between characters.
> 
> For instance, 0-127 (or 0x0000-0x007f) denotes the full ASCII range. As another example, ' '-'~' is equivalent to 32-127 and denotes the range of printable (visible) ASCII characters.

The problem with the current algorithm is that it splits the range on dashes before each character is parsed as a codepoint:
https://github.com/godotengine/godot/blob/d31794c4a26e5e10fc30c34a1ae9722fd9f50123/editor/import/resource_importer_imagefont.cpp#L133-L148
https://github.com/godotengine/godot/blob/d31794c4a26e5e10fc30c34a1ae9722fd9f50123/editor/import/resource_importer_imagefont.cpp#L74-L90

### Steps to reproduce

1. Create a new 2D project
2. Create an image of a dash character and place it in the root of the project directory
3. Import this image as `Font Data (Monospace Image Font)`
4. Add a character range of `'-'`
5. Click `Reimport (*)`
6. Observe in the Output window that the following is printed:
```
editor/import/resource_importer_imagefont.cpp:136 - Invalid range: "'-'"
```
7. Create a label with the text `-`
8. Set the font to the dash image
9. Observe that the dash is not rendered correctly and that there is an error in the scene tree:

image

### Minimal reproduction project

[Dash Bug Demo.zip](https://github.com/godotengine/godot/files/12823430/Dash.Bug.Demo.zip)

### PR DESCRIPTION
**Pull request by k0T0z**
Fixes #82880.

### CODE REVIEW

**Review by bruvzg — CHANGES_REQUESTED** [review]
This won't work and will break parsing of every range in ` U+xxxx` or hex form.

File: editor/import/resource_importer_imagefont.cpp [hunk 1]
```diff
@@ -130,6 +130,9 @@ Error ResourceImporterImageFont::import(const String &p_source_file, const Strin
 	int pos = 0;
 	for (int i = 0; i < ranges.size(); i++) {
 		int32_t start, end;
+		if (_decode_range(ranges[i], start)) {
+			continue;
+		}
 		Vector<String> tokens = ranges[i].split("-");
```
[the remark this answers is not in the document]
Comment (k0T0z) [reply editor/import/resource_importer_imagefont.cpp:136]: Very good, this looks much better, but can you give an example of a range of hex ``U+xxxx``?

[the remark this answers is not in the document]
Comment (k0T0z) [reply editor/import/resource_importer_imagefont.cpp:136]: @bruvzg What about this: ``'''-'''`` or ``'"'-'"'``?

**Comment by k0T0z** [comment]
@bruvzg I think this will do it. I added a step before splitting the range to replace every ``"'-'"`` with ``"0x2D"`` if existed and valid of course. I tested it and it working great.

File: editor/import/resource_importer_imagefont.h [hunk 2]
```diff
@@ -53,6 +53,9 @@ class ResourceImporterImageFont : public ResourceImporter {
 	virtual Error import(const String &p_source_file, const String &p_save_path, const HashMap<StringName, Variant> &p_options, List<String> *r_platform_variants, List<String> *r_gen_files = nullptr, Variant *r_metadata = nullptr) override;
 
 	ResourceImporterImageFont();
+
+private:
+	String CorrectOutlierSplitters(const String &range);
```
Comment (AThousandShips) [inline editor/import/resource_importer_imagefont.h:58]:
```suggestion
	String correct_outlier_splitters(const String &p_range);
```
Prefix arguments with `p_` and use `snake_case`

File: editor/import/resource_importer_imagefont.cpp [hunk 3]
```diff
@@ -89,6 +89,84 @@ bool ResourceImporterImageFont::_decode_range(const String &p_token, int32_t &r_
...
+			default:
+				if (nextToken[0] == tokens::character || nextToken[1] == tokens::character) {
+					accumulator.clear();
+					new_range += range[ptr]; // Not outlier? add it.
+				} else {
+					new_range += accumulator;
+					new_range += range.substr(ptr);
+					return new_range;
+				}
+				accumulator += range[ptr];
```
Comment (AThousandShips) [inline editor/import/resource_importer_imagefont.cpp:162]:
```suggestion
String ResourceImporterImageFont::correct_outlier_splitters(const String &p_range) {
	String new_range, accumulator;
	if (p_range.length() == 0) {
		return new_range;
	}
	enum Tokens {
		OPEN_SINGLE_QUOTE,
		CHARACTER,
		CLOSE_SINGLE_QUOTE,
		SPLITTER,
		INVALID
	};
	Tokens next_token[2] = { OPEN_SINGLE_QUOTE, CHARACTER };
	int ptr = 0 ;
	bool opened_quote = false, outlier_splitter = false;
	while (true) {
		if (ptr >= p_range.length()) {
			break;
		}
		switch (p_range[ptr]) {
			case '\'':
				if (next_token[0] == OPEN_SINGLE_QUOTE || next_token[1] == OPEN_SINGLE_QUOTE) {
					next_token[0] = CHARACTER;
					next_token[1] = INVALID;
				} else if (next_token[0] == CHARACTER || next_token[1] == CHARACTER) {
					next_token[0] = CLOSE_SINGLE_QUOTE;
					next_token[1] = INVALID;
				} else if (next_token[0] == CLOSE_SINGLE_QUOTE || next_token[1] == CLOSE_SINGLE_QUOTE) {
					next_token[0] = SPLITTER;
					next_token[1] = INVALID;
					accumulator.clear();
					if (outlier_splitter) {
						outlier_splitter = false;
						new_range += "0x2D"; // Outlier? replace it.
					} else {
						new_range += p_range.substr(ptr - 2, 3);
					}
				} else {
					new_range += accumulator;
					new_range += p_range.substr(ptr);
					return new_range;
				}
				accumulator += p_range[ptr];
				break;
			case '-':
				if (next_token[0] == CHARACTER || next_token[1] == CHARACTER) {
					next_token[0] = CLOSE_SINGLE_QUOTE;
					next_token[1] = INVALID;
					outlier_splitter = true;
				} else if (next_token[0] == SPLITTER || next_token[1] == SPLITTER) {
					next_token[0] = OPEN_SINGLE_QUOTE;
					next_token[1] = CHARACTER;
					accumulator.clear();
					new_range += '-'; // Not outlier? add it.
				} else {
					new_range += accumulator;
					new_range += p_range.substr(ptr);
					return new_range;
				}
				accumulator += p_range[ptr];
				break;
			default:
				if (next_token[0] == CHARACTER || next_token[1] == CHARACTER) {
					accumulator.clear();
					new_range += p_range[ptr]; // Not outlier? add it.
				} else {
					new_range += accumulator;
					new_range += p_range.substr(ptr);
					return new_range;
				}
				accumulator += p_range[ptr];
```

File: editor/import/resource_importer_imagefont.cpp [hunk 4]
```diff
@@ -129,21 +207,29 @@ Error ResourceImporterImageFont::import(const String &p_source_file, const Strin
 
 	int pos = 0;
 	for (int i = 0; i < ranges.size(); i++) {
+		String range{ ResourceImporterImageFont::CorrectOutlierSplitters( ranges[i]) };
```
Comment (AThousandShips) [inline editor/import/resource_importer_imagefont.cpp:210]:
```suggestion
		String range = correct_outlier_splitters(ranges[i]);
```

File: editor/import/resource_importer_imagefont.cpp [hunk 5]
```diff
@@ -129,21 +207,29 @@ Error ResourceImporterImageFont::import(const String &p_source_file, const Strin
...
 		int32_t start, end;
-		Vector<String> tokens = ranges[i].split("-");
+		Vector<String> tokens = range.split("-");
+		if (tokens.size() == 2) {
+			if (tokens[0] == tokens[1]) {
+				String token{ tokens[0] };
```
Comment (AThousandShips) [inline editor/import/resource_importer_imagefont.cpp:215]:
```suggestion
				String token = tokens[0];
```

File: editor/import/resource_importer_imagefont.cpp [hunk 6]
```diff
@@ -89,6 +89,84 @@ bool ResourceImporterImageFont::_decode_range(const String &p_token, int32_t &r_
 	}
 }
 
+String ResourceImporterImageFont::correct_outlier_splitters(const String &p_range) {
+	String new_range, accumulator;
+	if (p_range.length() == 0) {
```
Comment (AThousandShips) [inline editor/import/resource_importer_imagefont.cpp:94]:
```suggestion
	if (p_range.is_empty()) {
```

**Comment by k0T0z** [comment]
I would like this PR to be part of Hacktoberfest. Thank you!

File: editor/import/resource_importer_imagefont.cpp [hunk 7]
```diff
@@ -129,21 +207,29 @@ Error ResourceImporterImageFont::import(const String &p_source_file, const Strin
...
 		int32_t start, end;
-		Vector<String> tokens = ranges[i].split("-");
+		Vector<String> tokens = range.split("-");
+		if (tokens.size() == 2) {
+			if (tokens[0] == tokens[1]) {
+				String token = tokens[0];
+				tokens.clear();
+				tokens.push_back(token);
+			}
+		}
```
Comment (AThousandShips) [inline editor/import/resource_importer_imagefont.cpp:219]:
```suggestion
		if (tokens.size() == 2 && tokens[0] == tokens[1]) {
			String token = tokens[0];
			tokens.clear();
			tokens.push_back(token);
		}
```

**Comment by k0T0z** [comment]
@AThousandShips I want to finalize this by adding unit tests but actually there is no place for ``editor/imports/`` inside ``tests/`` so I just add a new directory there or there is a specific place for them?

**Comment by AThousandShips** [comment]
Don't think we do unit tests for the editor builds right now, think that needs to be discussed and see if it's possible, and if this can be tested without certain features available

**Comment by k0T0z** [comment]
Alright then, I will discuss this on the related channel ASAP 👍.

**Comment by AThousandShips** [comment]
No rush, probably a complex topic that might not be worth waiting for with this PR but add later, though this PR will be for 4.3 anyway

**Comment by k0T0z** [comment]
Cool, I hope I can be here for a more complex topic but I am busy with other things now. I get it and I intend to achieve more challenging things here 🤝

**Comment by AThousandShips** [comment]
You removed the method in the header file, accident?

**Comment by k0T0z** [comment]
I am just squashing commit, I will fix this now

**Comment by akien-mga** [comment]
Superseded by #88088. Thanks for the contribution!

**Pull request closed without merging**